File: ios-app/OutAndAbout/Core/Repositories/ReelRepository.swift
import Foundation
import FirebaseFirestore

protocol ReelRepositoryProtocol {
    func listenToPlayerReelJob(roomId: String, playerId: String) -> AsyncStream<ReelJob?>
    func listenToGroupReelJob(roomId: String) -> AsyncStream<ReelJob?>
    func getAllReelJobs(roomId: String) async throws -> [ReelJob]
}

final class ReelRepository: ReelRepositoryProtocol {
    private let firestore = FirestoreService.shared

    func listenToPlayerReelJob(roomId: String, playerId: String) -> AsyncStream<ReelJob?> {
        AsyncStream { continuation in
            let query = self.firestore.db
                .collection("reelJobs")
                .whereField("roomId", isEqualTo: roomId)
                .whereField("playerId", isEqualTo: playerId)
                .whereField("type", isEqualTo: "personal")
                .limit(to: 1)

            let listener = query.addSnapshotListener { snapshot, error in
                if let error {
                    print("[ReelRepository] Listener error: \(error)")
                    continuation.yield(nil)
                    return
                }
                let job = snapshot?.documents.first.flatMap { try? $0.data(as: ReelJob.self) }
                continuation.yield(job)
            }
            continuation.onTermination = { _ in listener.remove() }
        }
    }

    func listenToGroupReelJob(roomId: String) -> AsyncStream<ReelJob?> {
        AsyncStream { continuation in
            let query = self.firestore.db
                .collection("reelJobs")
                .whereField("roomId", isEqualTo: roomId)
                .whereField("type", isEqualTo: "group")
                .limit(to: 1)

            let listener = query.addSnapshotListener { snapshot, error in
                if let error {
                    print("[ReelRepository] Group listener error: \(error)")
                    continuation.yield(nil)
                    return
                }
                let job = snapshot?.documents.first.flatMap { try? $0.data(as: ReelJob.self) }
                continuation.yield(job)
            }
            continuation.onTermination = { _ in listener.remove() }
        }
    }

    func getAllReelJobs(roomId: String) async throws -> [ReelJob] {
        let snap = try await firestore.db
            .collection("reelJobs")
            .whereField("roomId", isEqualTo: roomId)
            .getDocuments()
        return try snap.documents.map { try $0.data(as: ReelJob.self) }
    }
}

File: web-app/src/repositories/reelRepository.ts
import { collection, query, where } from "firebase/firestore";
import { db } from "../firebase";
import { subscribeCollection } from "../lib/firestore";
import type { ReelJob } from "../types";

export function subscribeToReelJobs(
  roomId: string,
  onData: (jobs: ReelJob[]) => void,
  onError?: (err: Error) => void
): () => void {
  const q = query(collection(db, "reelJobs"), where("roomId", "==", roomId));
  return subscribeCollection<ReelJob>(q, onData, onError);
}

export function subscribeToMyReelJob(
  roomId: string,
  playerId: string,
  onData: (job: ReelJob | null) => void,
  onError?: (err: Error) => void
): () => void {
  const q = query(
    collection(db, "reelJobs"),
    where("roomId", "==", roomId),
    where("playerId", "==", playerId),
    where("type", "==", "personal")
  );
  return subscribeCollection<ReelJob>(q, (jobs) => onData(jobs[0] ?? null), onError);
}

export function subscribeToGroupReelJob(
  roomId: string,
  onData: (job: ReelJob | null) => void,
  onError?: (err: Error) => void
): () => void {
  const q = query(
    collection(db, "reelJobs"),
    where("roomId", "==", roomId),
    where("type", "==", "group")
  );
  return subscribeCollection<ReelJob>(q, (jobs) => onData(jobs[0] ?? null), onError);
}
